merkletree/merkletree_test.go

```diff
@@ -0,0 +1,403 @@
+	tree := New(sha256.New())
+	root := tree.Root()
+	if root != nil {
+		t.Error("root of empty tree should be nil")
+	}
+	_, proof, _, _ := tree.Prove()
+	if proof != nil {
+		t.Error("proof of empty tree should be nil")
+	}
+
+	// Get the proof of a tree that hasn't reached it's index.
+	err := tree.SetIndex(3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	tree.Push([]byte{1})
+	_, proof, _, _ = tree.Prove()
+	if proof != nil {
+		t.Fatal(err)
+	}
+	err = tree.SetIndex(2)
+	if err == nil {
+		t.Error("expecting error, shouldn't be able to reset a tree after pushing")
+	}
+
+	// Try nil values in VerifyProof.
+	mt := CreateMerkleTester(t)
+	if VerifyProof(sha256.New(), nil, mt.proveSets[1][0], 0, 1) {
+		t.Error("VerifyProof should return false for nil merkle root")
+	}
+	if VerifyProof(sha256.New(), []byte{1}, nil, 0, 1) {
+		t.Error("VerifyProof should return false for nil prove set")
+	}
+	if VerifyProof(sha256.New(), mt.roots[15], mt.proveSets[15][3][1:], 3, 15) {
+		t.Error("VerifyPRoof should return false for too-short prove set")
+	}
+	if VerifyProof(sha256.New(), mt.roots[15], mt.proveSets[15][10][1:], 10, 15) {
+		t.Error("VerifyPRoof should return false for too-short prove set")
+	}
+}
+
+// TestCompatibility runs BuildProof for a large set of trees, and checks that
+// verify affirms each proof, while rejecting for all other indexes (this
+// second half requires that all input data be unique).
+func TestCompatibility(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
+
+	// Brute force all trees up to size 'max'. Running time for this test is max^3.
+	max := 32
+	tree := New(sha256.New())
+	for i := 1; i < max; i++ {
+		// Try with proof at every possible index.
+		for j := 0; j < i; j++ {
+			// Push unique data into the tree.
+			tree.Reset()
+			err := tree.SetIndex(j)
+			if err != nil {
+				t.Fatal(err)
+			}
+			for k := 0; k < i; k++ {
+				tree.Push([]byte{byte(k)})
+			}
+
+			// Build the proof for the tree and run it through verify.
+			merkleRoot, proveSet, proveIndex, numLeaves := tree.Prove()
+			if !VerifyProof(sha256.New(), merkleRoot, proveSet, proveIndex, numLeaves) {
+				t.Error("proof didn't verify for indices", i, j)
+			}
+
+			// Check that verification fails for all other indices.
+			for k := 0; k < i; k++ {
+				if k == j {
+					continue
+				}
+				if VerifyProof(sha256.New(), merkleRoot, proveSet, k, numLeaves) {
+					t.Error("proof verified for indices", i, j, k)
+				}
+			}
+		}
+	}
+}
+
+// BenchmarkSha256_4MB runs a benchmark on hashing 4MB using sha256.
+func BenchmarkSha256_4MB(b *testing.B) {
+	data := make([]byte, 4*1024*1024)
+	rand.Read(data)
+	for i := 0; i < b.N; i++ {
+		sha256.Sum256(data)
+	}
+}
+
+// BenchmarkTree64Sha256 runs a benchmark on creating a Merkle tree out of 4MB
+// using 64 bytes at a time, using sha256 as the hashing algorithm.
+func BenchmarkTree64Sha256(b *testing.B) {
+	data := make([]byte, 4*1024*1024)
+	rand.Read(data)
+	segmentSize := 64
+	tree := New(sha256.New())
```

Comment:
these benchmarks should not include the time required to generate the data. Use `b.ResetTimer()`